### ISSUE TITLE
Do EDL full backups also contain the encrypted secrets from Trusted Execution Environments?

### ISSUE THREAD

**Issue by humdogm** [bkerler/edl#155]
Curious to know whether doing a full backup with edl.py rl, factory resetting the device, and restoring the backups will result in a device working exactly the way it was before, with userdata decryption working and everything?

I am also planning to reset my Oneplus 9 TMO edition with MSMDownloadtool, would it be useful to you if I did a USB packet capture of the entire process using Wireshark/USBPcap? I can upload to google drive.

**Comment by bkerler** [comment]
Yes, but I strongly recommend to use rf command instead for each lun instead.

And yes, a pcap log would help a lot to sort out possible issues with the op9 ;)

**Comment by humdogm** [comment]
Getting some weird errors when using rf,

```
Reading from physical partition 1, sector 0, sectors 2076
Progress: |██████████████████████████████████████████████████| 100.0% Completetor 2072 of 2076) 8.24 MB/sss
firehose
firehose - [LIB]: Error:
...
Reading from physical partition 2, sector 0, sectors 2076
Progress: |██████████████████████████████████████████████████| 100.0% Completetor 2072 of 2076) 31.43 MB/ss
firehose
firehose - [LIB]: Error:
...
Dumping sector 0 with sector count 8220 as UFS_lun3.bin.
firehose -
Reading from physical partition 3, sector 0, sectors 8220
Progress: |██████████████████████████████████████████████████| 100.0% Completector 8216 of 8220) 8.45 MB/ss
firehose
firehose - [LIB]: Error:
```
Below is the command I am using:
`python.exe C:\Users\humziq\git\edl\edl.py rf UFS_lun2.bin --memory=ufs --lun=2 --loader=C:/Users/humziq/git/edl/Loaders/oneplus/0000000000514d67_a26bc25799770106_fhprg_op9.bin`

**Comment by humdogm** [comment]
```
Reading from physical partition 4, sector 0, sectors 1572892
Progress: |██████████████████████████████████████████████████| 100.0% Read (Sector 1572816 of 1572892) 14.33 MB/s
Progress: |██████████████████████████████████████████████████| 100.0% Complete
firehose
firehose - [LIB]: Error:
```
Lun 4 also did it
When I use `tail -c $((4096*80)) UFS_lun4.bin | xxd -a | less`, I see that there is data in those sectors, should I be concerned about missing sectors? In the other LUNs (except for 0, haven't gotten that one yet) the sector before is all zeros and the sectors after are also all zeros. Is it just not updating the terminal that it's reading the final sectors?

**Comment by bkerler** [comment]
As long as the files contain data and percentage is 100%, all should be good. There is a small bug with overread of data because of backup gpt, but even if backup gpt would be missing, it's no problem as the main gpt is still intact.

**Comment by humdogm** [comment]
Would it also be ok to flash the backed up partitions using fastboot after factory reset and unlocking? I will check the LUNs for data in between GPT partitions.

**Comment by bkerler** [comment]
If you backup a locked device and flash on a unlocked device, the device won't boot as the crypto keys are changed on unlocking.

**Comment by humdogm** [comment]
I only ask because my OP9 TMO doesn't seem to have write access through EDL mode. I attempt to write data to it but upon reading it back nothing changes. Though I have seen others talk about writing to their OP9s.

Factory resetting doesn't erase crypto keys?

**Comment by humdogm** [comment]
Here's the log.txt of trying to write to vbmeta_a
[log.txt](https://github.com/bkerler/edl/files/7230318/log.txt)

**Comment by humdogm** [comment]
Here's the settings.xml file that comes with the firmware if that helps with anything
[settings.xml.txt](https://github.com/bkerler/edl/files/7233166/settings.xml.txt)

**Comment by bkerler** [comment]
Factory resetting of course does change the crypto keys. The main issue is that the challenge vs. response doesn't work as it seems. I'd need a log of the start of the flashing with MSMDownloadTool (pcap) to fix it.

**Comment by humdogm** [comment]
Here's the beginning of flashing with MSMDownloadTool:
https://drive.google.com/file/d/1xIdDv-eDdBsGUpgqyK_Z6KqvT_yugRT_/view?usp=sharing

**Comment by humdogm** [comment]
Lmk if you want me to test anything

**Comment by humdogm** [comment]
I managed to get it working by changing line 945 of firehose.py to `self.supported_functions = ["demacia", "setprojmodel"]` and in oneplus.py by changing part of the class oneplus1's init method to
```
self.random_postfix = "0iyFR00pPnoqjVNL"
self.Version = "lemonadet_22_O.01_210307"
```
With this I managed to write then read to verify the write went through.
self.Version can be found in param.bin at offset 0x3038 to 0x30ff on my particular oneplus device, not sure about others.
The random postfix will probably have to come from leaked msmdownloadtools I'm afraid.

**Comment by bkerler** [comment]
Thanks for the details, I will have a closer look how it can be improved.

**Comment by bkerler** [comment]
Thanks to your input, commit 91c0f93 should work fine now.
